lte/gateway/c/session_manager/SessionStore.h

```diff
@@ -26,15 +27,29 @@
 
 namespace magma {
 namespace lte {
+using std::experimental::optional;
 
-typedef std::unordered_map<
-    std::string, std::vector<std::unique_ptr<SessionState>>>
-    SessionMap;
 // Value int represents the request numbers needed for requests to PCRF
 typedef std::set<std::string> SessionRead;
 typedef std::unordered_map<
     std::string, std::unordered_map<std::string, SessionStateUpdateCriteria>>
     SessionUpdate;
+
+enum SessionSearchCriteriaType {
+  IMSI_AND_APN        = 0,
+  IMSI_AND_SESSION_ID = 1,
+};
+
+struct SessionSearchCriteria {
+  std::string imsi;
+  SessionSearchCriteriaType search_type;
+  std::string secondary_key;
+  SessionSearchCriteria(
+      const std::string p_imsi, SessionSearchCriteriaType p_type,
+      const std::string p_secondary_key)
+      : imsi(p_imsi), search_type(p_type), secondary_key(p_secondary_key) {}
+};
+
```

Comment:
This is opinion and you don't need to change it:

isn't this struct the same as having two separate functions?
```
SessionStore::find_session_by_imsi_and_apn(imsi, apn)
SessionStore::find_session_by_imsi_and_sessionId(imsi, sessionID)
```

Would it not be better just have 2 separate functions to make the API simpler. The addition of the struct may add some complexity 


(code below is copied pasted so may have errors!!)
```
optional<SessionVector::iterator> SessionStore::find_session_by_imsi_and_apn(
    SessionMap& session_map, SessionSearchCriteria criteria) {
  auto it = session_map.find(criteria.imsi);
  if (it == session_map.end()) {
    return {};
  }
  auto& sessions = it->second;
  for (auto it = sessions.begin(); it != sessions.end(); ++it) {
        if ((*it)->get_session_id() == criteria.secondary_key) {
          return it;
        }
  }
  return {};
}


optional<SessionVector::iterator> SessionStore::find_session_by_imsi_and_sessionId(
    SessionMap& session_map, SessionSearchCriteria criteria) {
  auto it = session_map.find(criteria.imsi);
  if (it == session_map.end()) {
    return {};
  }
  auto& sessions = it->second;
  for (auto it = sessions.begin(); it != sessions.end(); ++it) {
        if ((*it)->get_config().common_context.apn() ==
            criteria.secondary_key) {
          return it;
        }
    
  }
  return {};
}
```

If you are concerned about code reusability, maybe you can do something similar of what you did without having to expose that struct. But my opinion is that since you are just saving a `for` i would prioritize readability vs code reusability here

Reply:
Yeah, so I made this choice because I am anticipating a lot more keys to search by. For example, one use case I have upcoming is to match by IP address. So instead of creating a whole new function, I thought it would be simpler to just add a new type and a few lines of logic.

Reply:
Sounds good then. 

Maybe we can make it prettier once we have more keys